File: src/test/java/org/tindalos/principle/domain/checker/BlueprintTest.java
package org.tindalos.principle.domain.checker;

import static org.junit.Assert.assertEquals;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.junit.BeforeClass;
import org.junit.Test;
import org.tindalos.principle.domain.core.DesignQualityCheckConfiguration;
import org.tindalos.principle.domain.core.Package;
import org.tindalos.principle.domain.coredetector.CheckResult;
import org.tindalos.principle.domain.detector.submodulesblueprint.Submodule;
import org.tindalos.principle.domain.detector.submodulesblueprint.SubmoduleId;
import org.tindalos.principle.domain.detector.submodulesblueprint.SubmodulesBlueprintCheckResult;
import org.tindalos.principle.domain.expectations.DesignQualityExpectations;
import org.tindalos.principle.domain.util.ListConverter;
import org.tindalos.principle.domain.expectations.SubmodulesBlueprint;
import org.tindalos.principle.infrastructure.di.PoorMansDIContainer;
import org.tindalos.principle.infrastructure.plugin.Checks;

import com.google.common.collect.Maps;
import com.google.common.collect.Sets;

public class BlueprintTest {

    @BeforeClass
    public static void setup() {
        TestFixture.setLogger();
    }

    @Test
    public void missingAndIllegal() {
        
        SubmodulesBlueprintCheckResult result = run("org.tindalos.principletest.submodulesblueprint","src/test/resources/principle_blueprint_test.yaml");
        
        Map<Submodule, Set<Submodule>> illegalDependencies = ListConverter.convertMapSet(result.illegalDependencies());
        Map<Submodule, Set<Submodule>> missingDependencies = ListConverter.convertMapSet(result.missingDependencies());
        
        Submodule mod1 = fakeSubmodule("MOD1");
        Submodule mod2 = fakeSubmodule("MOD2");
        Submodule mod3 = fakeSubmodule("MOD3");
        
        Map<Submodule, Set<Submodule>> expectedIllegalDependencies = Maps.newHashMap();
        expectedIllegalDependencies.put(mod3, Sets.newHashSet(mod2));

        Map<Submodule, Set<Submodule>> expectedMissingDependencies = Maps.newHashMap();
        expectedMissingDependencies.put(mod1, Sets.newHashSet(mod2));
        
        assertEquals(expectedIllegalDependencies, illegalDependencies);
        assertEquals(expectedMissingDependencies, missingDependencies);
    }
    
    @Test
    public void overlapping() {
    	SubmodulesBlueprintCheckResult result = run("org.tindalos.principletest.submodulesblueprint","src/test/resources/principle_blueprint_test_overlapping.yaml");
    	System.out.println(result);
    }
    
    private void assertDependencies(Map<Submodule, Set<Submodule>> actualDependencies, Submodule referer, Submodule... referees) {
		Map<Submodule, Set<Submodule>> expected = Maps.newHashMap();
		assertEquals(actualDependencies, expected);
	}

	private static Submodule fakeSubmodule(String name) {
    	Set<Package> empty = Sets.newHashSet();
		return new Submodule(new SubmoduleId(name), ListConverter.convert(empty), ListConverter.convert((new HashSet<SubmoduleId>())));
    }
    
       
    private SubmodulesBlueprintCheckResult run(String basePackage, String location) {
        DesignQualityExpectations checks = prepareChecks(location);
        DesignQualityCheckService designQualityCheckService = PoorMansDIContainer.getDesignCheckService(basePackage);
        DesignQualityCheckResults checkResults = designQualityCheckService.analyze(new DesignQualityCheckConfiguration(checks, basePackage));
        List<CheckResult> resultList = checkResults.resultList();
        assertEquals(1, resultList.size());
        return (SubmodulesBlueprintCheckResult) resultList.get(0);
    }

    private static DesignQualityExpectations prepareChecks(String location) {
        Checks checks = new Checks();
        checks.submodulesBlueprint_$eq(submodulesBlueprint(location));
        return checks;
    }

	private static SubmodulesBlueprint submodulesBlueprint(String location) {
		return new SubmodulesBlueprint(location, 0);
	}

}

File: src/test/scala/org/tindalos/principle/domain/checker/BlueprintTest.scala
package org.tindalos.principle.domain.checker

import org.junit.Assert.assertEquals
import org.junit._
import org.tindalos.principle.domain.AnalysisInput
import org.tindalos.principle.domain.core.AnalysisPlan
import org.tindalos.principle.domain.analyzers.submodulesblueprint._
import org.tindalos.principle.domain.constraints._
import org.tindalos.principle.domain.core.packages.PackageWithMetrics

import scala.collection.JavaConverters._
import org.tindalos.principle.infrastructure.JDependBasedPackageListBuilder
import org.tindalos.principle.infrastructure.analyzers.submodulesblueprint.YAMLBasedSubmodulesBlueprintProvider
import org.tindalos.principle.infrastructure.di.PoorMansDIContainer

class BlueprintTest {

  @Before
  def setup() = {
    TestFixture.setLogger()
  }

  @Test
  def missingAndIllegal() {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_test.yaml")

    val mod1 = fakeSubmodule("MOD1")
    val mod2 = fakeSubmodule("MOD2")
    val mod3 = fakeSubmodule("MOD3")

    assertEquals(java.util.Map.of(mod3, java.util.Set.of[Submodule](mod2)), result.illegalDependencies())
    assertEquals(java.util.Map.of(mod1, java.util.Set.of[Submodule](mod2)), result.missingDependencies())
  }

  @Test
  def overlapping() = {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_test_overlapping.yaml")

    assertEquals(java.util.Map.of(), result.illegalDependencies())
    assertEquals(java.util.Map.of(), result.missingDependencies())
    assert(!result.overlaps().isEmpty, "Expected overlaps to be detected")
  }

  @Test
  def violationsCount() = {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_test.yaml")

    assertEquals(2, result.violationsNumber())
    assertEquals(1, result.illegalDependencies().size)
    assertEquals(1, result.missingDependencies().size)
  }

  @Test
  def illegalDependenciesOnly() = {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_test.yaml")

    val mod3 = fakeSubmodule("MOD3")
    val mod2 = fakeSubmodule("MOD2")

    assert(result.illegalDependencies().containsKey(mod3), "MOD3 should have illegal dependencies")
    val illegalDeps = result.illegalDependencies().get(mod3)
    assert(illegalDeps.contains(mod2), "MOD3 illegally depends on MOD2")
  }

  @Test
  def missingDependenciesOnly() = {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_test.yaml")

    val mod1 = fakeSubmodule("MOD1")
    val mod2 = fakeSubmodule("MOD2")

    assert(result.missingDependencies().containsKey(mod1), "MOD1 should have missing dependencies")
    val missingDeps = result.missingDependencies().get(mod1)
    assert(missingDeps.contains(mod2), "MOD1 is missing dependency on MOD2")
  }

  @Test
  def expectationsFailed_whenViolationsExceedThreshold() = {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_test.yaml")
    assert(result.constraintViolated(), "Expectations should fail when violations exceed threshold")
  }

  @Test
  def verifySubmoduleEquality() = {
    val mod1a = fakeSubmodule("MOD1")
    val mod1b = fakeSubmodule("MOD1")
    val mod2 = fakeSubmodule("MOD2")

    // Submodules with same ID should be equal
    assertEquals(mod1a, mod1b)
    assert(mod1a != mod2, "Submodules with different IDs should not be equal")
  }

  @Test
  def verifyResultStructure() = {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_test.yaml")

    assertEquals(0, result.threshold())
    assert(result.illegalDependencies().isInstanceOf[java.util.Map[_, _]], "Illegal dependencies should be a Map")
    assert(result.missingDependencies().isInstanceOf[java.util.Map[_, _]], "Missing dependencies should be a Map")
  }

  @Test
  def blueprintOk_parsingSucceeds() = {
    val result = run("org.tindalos.principletest.submodulesblueprint", "src/test/resources/principle_blueprint_ok.yaml")

    assertEquals(0, result.threshold())
    assert(result.overlaps().isEmpty, "Valid blueprint should have no overlaps")
    assert(result.violationsNumber() >= 0, "Violations should be non-negative")
  }

  def fakeSubmodule(name: String) = {
    new Submodule(new SubmoduleId(name), java.util.Set.of[PackageWithMetrics](), java.util.Set.of[SubmoduleId]())
  }

  private def run(basePackage: String, location: String) = {
    val provider = new YAMLBasedSubmodulesBlueprintProvider()
    val submoduleDefinitions = provider.readSubmoduleDefinitions(basePackage, location, 0)
    val constraints = Constraints.builder().submoduleDefinitions(submoduleDefinitions).build()
    val packageListProducer = new JDependBasedPackageListBuilder(basePackage)
    val packageList = packageListProducer.build()
    val analysisRunner= PoorMansDIContainer.buildAnalysisRunner()
    val plan = new AnalysisPlan(constraints, basePackage)
    val result = analysisRunner.run(new AnalysisInput(packageList.map(p => p: PackageWithMetrics).asJava, Set.empty.asJava, plan))
    assertEquals(1, result.length)
    result.head.asInstanceOf[SubmodulesBlueprintAnalysisResult]
  }


}